File: src/main/java/com/example/chatService/service/DMService.java

package com.example.chatService.service;

import com.example.chatService.dto.DMRoomDto;
import com.example.chatService.entity.DMMessage;
import com.example.chatService.entity.DMRoom;
import com.example.chatService.entity.User;
import com.example.chatService.repository.DMMessageRepository;
import com.example.chatService.repository.DMRoomRepository;
import com.example.chatService.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.UUID;

@Service
@Transactional
@RequiredArgsConstructor
public class DMService {

    private final DMRoomRepository roomRepository;
    private final DMMessageRepository dmMessageRepository;
    private final UserRepository userRepository;

    public DMRoom startOrGetRoom(String userA, String userB) {
        return roomRepository.findByUserAIdAndUserBId(userA, userB)
                .or(() -> roomRepository.findByUserBIdAndUserAId(userA, userB))
                .orElseGet(() -> {
                    DMRoom room = DMRoom.builder()
                            .roomId(UUID.randomUUID().toString())
                            .userAId(userA)
                            .userBId(userB)
                            .build();
                    return roomRepository.save(room);
                });
    }

    public List<DMMessage> getMessages(String roomId) {
        DMRoom room = roomRepository.findById(roomId)
                .orElseThrow(() -> new RuntimeException("Room not found"));
        return dmMessageRepository.findByRoomOrderBySentAtAsc(room);
    }

    public DMMessage sendMessage(String roomId, String senderId, String content) {
        DMRoom room = roomRepository.findById(roomId)
                .orElseThrow(() -> new RuntimeException("Room not found"));

        DMMessage msg = DMMessage.builder()
                .room(room)
                .senderId(senderId)
                .content(content)
                .sentAt(OffsetDateTime.now())
                .isRead(false)
                .build();

        room.setLastMessageTime(OffsetDateTime.now());
        roomRepository.save(room);

        return dmMessageRepository.save(msg);
    }

    public String getReceiverId(String roomId, String senderId) {
        DMRoom room = roomRepository.findById(roomId)
                .orElseThrow(() -> new IllegalArgumentException("Room not found"));

        return room.getUserAId().equals(senderId)
                ? room.getUserBId()
                : room.getUserAId();
    }
    public List<DMRoomDto> getUserRoomsWithUnread(String myUserId) {
        List<DMRoom> rooms = roomRepository.findByUser(myUserId);
        return rooms
                .stream()
                .map(room -> {
                    String targetUserId = room.getUserAId().equals(myUserId) ? room.getUserBId() : room.getUserAId();

            User targetUser = userRepository.findById(targetUserId)
                    .orElseThrow(() -> new RuntimeException("User not found"));

            int unread = dmMessageRepository.countUnread(room.getRoomId(), myUserId);

            return new DMRoomDto(
                    room.getRoomId(),
                    targetUser.getId(),
                    targetUser.getUsername(),
                    unread
            );
        }).toList();
    }

    @Transactional
    public void markMessagesAsRead(String roomId, String userId) {
        dmMessageRepository.markAsReadByRoomAndReceiver(roomId, userId);
    }
}

File: src/main/java/com/example/chatService/service/DMService.kt
package com.example.chatService.service

import com.example.chatService.dto.DMRoomDto
import com.example.chatService.entity.DMMessage
import com.example.chatService.entity.DMRoom
import com.example.chatService.repository.DMMessageRepository
import com.example.chatService.repository.DMRoomRepository
import com.example.chatService.repository.UserRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.OffsetDateTime
import java.util.UUID

@Service
@Transactional
class DMService(
        private val roomRepository : DMRoomRepository,
        private val dmMessageRepository : DMMessageRepository,
        private val userRepository : UserRepository
) {
    fun startOrGetRoom(userA: String, userB : String) : DMRoom {
        return roomRepository.findByUserAIdAndUserBId(userA, userB)
                .or { roomRepository.findByUserBIdAndUserAId(userA, userB) }
                .orElseGet {
                    val room = DMRoom(
                            roomId = UUID.randomUUID().toString(),
                            userAId = userA,
                            userBId = userB
                    )
                    roomRepository.save(room);
                }
    }
    fun getMessages(roomId : String) : List<DMMessage> {
        val room = roomRepository.findById(roomId)
                .orElseThrow { RuntimeException("Room not found") }
        return dmMessageRepository.findByRoomOrderBySentAtAsc(room);
    }

    fun sendMessage(roomId: String, senderId: String, content : String) : DMMessage {
        val room = roomRepository.findById(roomId)
                .orElseThrow { RuntimeException("Room not found") }

        val msg = DMMessage(
                room = room,
                senderId = senderId,
                content = content,
                sentAt = OffsetDateTime.now(),
                isRead = false
        )
        room.lastMessageTime = OffsetDateTime.now()
        roomRepository.save(room)

        return dmMessageRepository.save(msg)
    }

    fun getReceiverId(roomId : String, senderId : String) : String? {
        val room = roomRepository.findById(roomId)
                .orElseThrow { IllegalArgumentException("Room not found") }

        return if (room.userAId == senderId) {
            room.userBId
        } else {
            room.userAId
        }
    }

    fun getUserRoomsWithUnread(myUserId : String) : List<DMRoomDto> {
        val rooms = roomRepository.findByUser(myUserId)
        return rooms.map { room ->
            val targetUserId = if (room.userAId == myUserId) room.userBId else room.userAId

            val targetUser = userRepository.findById(targetUserId ?: throw RuntimeException("user not found"))
                    .orElseThrow { RuntimeException("user not found") }
            val unread = dmMessageRepository.countUnread(
                    room.roomId ?: throw RuntimeException("room id not found"),
                    myUserId
            )

            DMRoomDto(
                    room.roomId,
                    targetUser.id,
                    targetUser.username,
                    unread)
        }
    }
    @Transactional
    fun markMessagesAsRead(roomId : String, userId: String) {
        dmMessageRepository.markAsReadByRoomAndReceiver(roomId, userId)
    }
}
